lib/launcher.rb

```diff
@@ -66,39 +74,63 @@ def https_app(https_port, plugins = https_plugins)
-      ssl_options |= OpenSSL::SSL::OP_NO_TLSv1 if defined?(OpenSSL::SSL::OP_NO_TLSv1)
-
-      Proxy::SETTINGS.tls_disabled_versions&.each do |version|
-        constant = OpenSSL::SSL.const_get("OP_NO_TLSv#{version.to_s.tr('.', '_')}") rescue nil
-
-        if constant
-          logger.info "TLSv#{version} will be disabled."
-          ssl_options |= constant
-        else
-          logger.warn "TLSv#{version} was not found."
-        end
-      end
-
-      {
+      app_details = {
         :app => app,
-        :server => :webrick,
+        :server => settings.http_server_type,
         :DoNotListen => true,
         :Port => https_port, # only being used to correctly log https port being used
         :Logger => ::Proxy::LogBuffer::Decorator.instance,
         :ServerSoftware => "foreman-proxy/#{Proxy::VERSION}",
-        :SSLEnable => true,
-        :SSLVerifyClient => OpenSSL::SSL::VERIFY_PEER,
-        :SSLPrivateKey => load_ssl_private_key(settings.ssl_private_key),
-        :SSLCertificate => load_ssl_certificate(settings.ssl_certificate),
-        :SSLCACertificateFile => settings.ssl_ca_file,
-        :SSLOptions => ssl_options,
         :daemonize => false,
       }
+      case settings.http_server_type
+      when "webrick"
+        ssl_options = OpenSSL::SSL::SSLContext::DEFAULT_PARAMS[:options]
+        ssl_options |= OpenSSL::SSL::OP_CIPHER_SERVER_PREFERENCE if defined?(OpenSSL::SSL::OP_CIPHER_SERVER_PREFERENCE)
+        ssl_options |= OpenSSL::SSL::OP_NO_SSLv2 if defined?(OpenSSL::SSL::OP_NO_SSLv2)
+        ssl_options |= OpenSSL::SSL::OP_NO_SSLv3 if defined?(OpenSSL::SSL::OP_NO_SSLv3)
+        ssl_options |= OpenSSL::SSL::OP_NO_TLSv1 if defined?(OpenSSL::SSL::OP_NO_TLSv1)
+
+        if settings.tls_disabled_versions
+          settings.tls_disabled_versions&.each do |version|
+            constant = OpenSSL::SSL.const_get("OP_NO_TLSv#{version.to_s.tr('.', '_')}") rescue nil
+
+            if constant
+              logger.info "TLSv#{version} will be disabled."
+              ssl_options |= constant
+            else
+              logger.warn "TLSv#{version} was not found."
+            end
+          end
+        end
+
+        app_details[:SSLEnable] = true
+        app_details[:SSLVerifyClient] = OpenSSL::SSL::VERIFY_PEER
+        app_details[:SSLCACertificateFile] = settings.ssl_ca_file
+        app_details[:SSLPrivateKey] = load_ssl_private_key(settings.ssl_private_key)
+        app_details[:SSLCertificate] = load_ssl_certificate(settings.ssl_certificate)
+        app_details[:SSLOptions] = ssl_options
+      when "puma"
+        # https://github.com/puma/puma#binding-tcp--sockets
+        app_details[:SSLArgs] = {
+          :ca => settings.ssl_ca_file,
+          :key => settings.ssl_private_key,
+          :cert => settings.ssl_certificate,
+          :verify_mode => 'peer',
+        }
+        app_details[:SSLArgs][:no_tlsv1] = "true" if settings.tls_disabled_versions.include?("1.0")
```

Comment:
For webrick 1.0 is always disabled due to security concerns.
```suggestion
        app_details[:SSLArgs][:no_tlsv1] = "true"
```

Reply:
Yes, it's disabled even for Puma in `lib/proxy/settings/global.rb`. I am keeping this since we support enabling this via settings. Let's give power to our users but ship with a sane defaults.